### PR TITLE
Support Syndication Support Rss Optional Elements

### PR DESCRIPTION
**Pull request by shmao**
I'm prototyping supporting RSS optional elements using the `<docs>`element. Please let me know if you have any concerns or suggestions. I'll apply this to other optional elements after this PR.

The change enable users to read/wring `<docs>` via `SyndicationFeed.Documentation `. Here's a couple of things worth to note:

1. Any existing applications that read/write `<docs>` via `SyndicationFeed.ElementExtensions` would still work after the change.
2. If `SyndicationFeed.Documentation` is ever used and becomes not null, `SyndicationFeedFormatter.WriteTo` would write `<docs>` with the value of `SyndicationFeed.Documentation ` and ignore the corresponding ElementExtension.

Ref #25608.

### CODE REVIEW

File: src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs [hunk 1]
```diff
@@ -265,6 +269,44 @@ public TextSyndicationContent Title
...
+                                      .Where((e) => e.OuterName == Rss20Constants.DocumentationTag && e.OuterNamespace == Rss20Constants.Rss20Namespace)
+                                      .FirstOrDefault();
+
+            if (documentationElement == null)
+                return null;
+
+            using (XmlReader reader = documentationElement.GetReader())
+            {
+                SyndicationLink documentation = new Rss20FeedFormatter().ReadAlternateLink(reader, BaseUri);
+                return documentation;
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs]: ReadAlternateLink uses a UriParser which is why you need to instantiate an instance of Rss20FeedFormatter. This is just going to use the default Uri parser so we lose the benefit of the extensibility here. We should make ReadAlternateLink be a static method which takes a Uri parser as a parameter . Then do one of two things. 1) Have the class which instantiates the SyndicationFeed class set the parser in SyndicationFeed so any custom parser can be passed to ReadAlternateLink. 2) Pass the default Uri parser to ReadAlternateLink as that's what's being used here anyway.

Comment (shmao) [reply src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs]: Sounds good. I went with option 2).

File: src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs [hunk 2]
```diff
@@ -51,16 +51,20 @@ public static void SyndicationFeed_Load_Write_RSS_Feed()
...
 
                 // *** EXECUTE *** \\
                 //Write the same feed that was read.
-                XmlWriter xmlw = XmlWriter.Create(path);
-                Rss20FeedFormatter atomFeed = new Rss20FeedFormatter(sf);
-                atomFeed.WriteTo(xmlw);
-                xmlw.Close();
+                using (XmlWriter xmlw = XmlWriter.Create(path))
+                {
+                    Rss20FeedFormatter atomFeed = new Rss20FeedFormatter(sf);
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs]: This seems oddly written. An instance of Rss20FeedFormatter is called atomFeed when we have an atom formatter?

File: src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs [hunk 3]
```diff
@@ -51,16 +51,20 @@ public static void SyndicationFeed_Load_Write_RSS_Feed()
...
+                    atomFeed.WriteTo(xmlw);
+                }
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs]: There's no verification that we round trip all data?

File: src/System.ServiceModel.Syndication/tests/TestFeeds/rssOptionalElements.xml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      <description>Search Online</description>
+      <title>Search</title>
+      <asdasdasd>asdasdasdasdasd</asdasdasd>
+      <link>http://www.contoso.no/search?</link>
+      <name>input Name</name>
+    </textInput>
+    <!--
+    <image>
+      <url>http://2.bp.blogspot.com/-NA5Jb-64eUg/URx8CSdcj_I/AAAAAAAAAUo/eCx0irI0rq0/s1600/bg_Contoso_logo3-20120824073001907469-620x349.jpg</url>
+      <title>The title is not the same to the original one</title>
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/tests/TestFeeds/rssOptionalElements.xml]: This is commented out XML but this is a very real URL to a real image. Do we want to remove this or sanitize the values?

File: src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs (hunk 2 above)
Comment (shmao) [reply src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs]: Good catch. The naming error was there before I changed this. Here I'm fixing the usage of the `Close()` method which was not called when exception happened in `WriteTo`.

I'll fix the names.

File: src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs (hunk 3 above)
Comment (shmao) [reply src/System.ServiceModel.Syndication/tests/BasicScenarioTests.cs]: Yeah, I noticed this problem. But this is not a new added test. I'm fixing the usage of `Close` here.

File: src/System.ServiceModel.Syndication/tests/TestFeeds/rssOptionalElements.xml (hunk 4 above)
Comment (shmao) [reply src/System.ServiceModel.Syndication/tests/TestFeeds/rssOptionalElements.xml]: There's a bug with the original implementation of reading `<image>`, which read different values and failed the tests. I'm commenting out this section out for now. We will need to revisit this when we add support for `<image>`.

File: src/System.ServiceModel.Syndication/ref/System.ServiceModel.Syndication.netcoreapp.cs [hunk 5]
```diff
@@ -7,6 +7,10 @@
 
 namespace System.ServiceModel.Syndication
 {
+    public partial class SyndicationFeed
+    {
+        public SyndicationLink Documentation { get{ throw null; } set{ } }
+    }
```
Comment (stephentoub) [inline src/System.ServiceModel.Syndication/ref/System.ServiceModel.Syndication.netcoreapp.cs]: Has this been through API review?  If not, we should close the PR until it's been vetted.

Comment (shmao) [reply src/System.ServiceModel.Syndication/ref/System.ServiceModel.Syndication.netcoreapp.cs]: This has not been approved yet. I added the `NO MERGE` label on the PR for now. I'd like to get feedback from others and see if it's even possible or make sense to add the new APIs before sending out API request..

File: src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/Rss20FeedFormatter.cs [hunk 6]
```diff
@@ -1025,6 +1025,13 @@ private void WriteFeed(XmlWriter writer)
                 writer.WriteEndElement(); // image
             }
 
+            // Optional spec items
+            // docs
+            if (Feed.GetDocumentation() != null)
+            {
+                writer.WriteElementString(Rss20Constants.DocumentationTag, Feed.GetDocumentation().Uri.ToString());
+            }
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/Rss20FeedFormatter.cs]: Uri can be null

File: src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs [hunk 7]
```diff
@@ -265,6 +269,44 @@ public TextSyndicationContent Title
             set { _title = value; }
         }
 
+        internal SyndicationLink GetDocumentation()
+        {
```
Comment (mconnew) [inline src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs]: This is basically a property getter. Can you change this to a read only property with a name such as InternalDocumentation.

**Comment by mconnew** [comment]
Just the one small issue (method -> property), otherwise LGTM.

File: src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs [hunk 8]
```diff
@@ -265,6 +275,300 @@ public TextSyndicationContent Title
...
+        public int TimeToLive
+        {
+            get
+            {
+                if (_timeToLive == null)
+                {
+                    _timeToLive = TryReadTimeToLiveFromExtension(ElementExtensions);
+                }
+
+                return _timeToLive.Value;
```
Comment (JonHanna) [inline src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs]: Is `InvalidOperationException` if `TryReadTimeToLiveFromExtension` returns null appropriate?

Comment (shmao) [reply src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/SyndicationFeed.cs]: Good catch. I changed the type of the property to `int?`. Thanks.

**Comment by shmao** [comment]
@justinvp @JonHanna Thanks for reviewing the PR. I've made changes per your suggestions.

**Comment by shmao** [comment]
The proposed the APIs have not been reviewed yet. See https://github.com/dotnet/corefx/issues/25718.

I will be out of office and be back on Jan 2, 2018. I'm leaving the PR open for now. If the the APIs are approved, please feel free to merge the PR.  If anyone feels the PR being open for too long, please feel free to close it. Thanks.

**Comment by shmao** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by shmao** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by karelz** [comment]
Closing, it's replaced by PR #26612

**Pull request closed without merging**